File: app/src/main/java/com/slash/batterychargelimit/receivers/EnableWidgetIntentReceiver.java
package com.slash.batterychargelimit.receivers;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.widget.RemoteViews;
import android.widget.Toast;

import com.slash.batterychargelimit.R;
import com.slash.batterychargelimit.SharedMethods;
import eu.chainfire.libsuperuser.Shell;

import static com.slash.batterychargelimit.Constants.ENABLE;
import static com.slash.batterychargelimit.Constants.SETTINGS;
import static com.slash.batterychargelimit.Constants.INTENT_TOGGLE_ACTION;
import static com.slash.batterychargelimit.EnableWidget.buildButtonPendingIntent;
import static com.slash.batterychargelimit.EnableWidget.pushWidgetUpdate;

/**
 * Created by harsha on 5/5/17.
 */

public class EnableWidgetIntentReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
        if(intent.getAction().equals(INTENT_TOGGLE_ACTION)){
            SharedPreferences settings = context.getSharedPreferences(SETTINGS, 0);
            if (Shell.SU.available()) {
                boolean enable = !settings.getBoolean(ENABLE, false);
                settings.edit().putBoolean(ENABLE, enable).apply();
                if(enable) {
                    SharedMethods.startService(context);
                } else {
                    SharedMethods.stopService(context);
                }
                updateWidget(context, enable);
            } else {
                Toast.makeText(context, R.string.root_denied, Toast.LENGTH_LONG).show();
            }
        }
    }

    public static void updateWidget(Context context, boolean enable){
        RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.enable);

        remoteViews.setImageViewResource(R.id.enable, getImage(enable));
        remoteViews.setOnClickPendingIntent(R.id.enable, buildButtonPendingIntent(context));

        pushWidgetUpdate(context, remoteViews);
    }

    public static int getImage(boolean enabled){
        if (enabled) {
            return R.drawable.widget_enabled;
        } else {
            return R.drawable.widget_disabled;
        }
    }
}

File: app/src/main/java/com/slash/batterychargelimit/receivers/EnableWidgetIntentReceiver.kt
package com.slash.batterychargelimit.receivers

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import android.widget.Toast
import com.slash.batterychargelimit.Constants.CHARGE_LIMIT_ENABLED
import com.slash.batterychargelimit.Constants.INTENT_TOGGLE_ACTION
import com.slash.batterychargelimit.EnableWidget
import com.slash.batterychargelimit.R
import com.slash.batterychargelimit.Utils
import eu.chainfire.libsuperuser.Shell

class EnableWidgetIntentReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action == INTENT_TOGGLE_ACTION) {
            val settings = Utils.getSettings(context)
            if (Shell.SU.available()) {
                val enable = !settings.getBoolean(CHARGE_LIMIT_ENABLED, false)
                settings.edit().putBoolean(CHARGE_LIMIT_ENABLED, enable).apply()
                if (enable) {
                    Utils.startServiceIfLimitEnabled(context)
                } else {
                    Utils.stopService(context)
                }
                updateWidget(context, enable)
            } else {
                Toast.makeText(context, R.string.root_denied, Toast.LENGTH_LONG).show()
            }
        }
    }

    companion object {

        fun updateWidget(context: Context, enable: Boolean) {
            val remoteViews = RemoteViews(context.packageName, R.layout.widget_button)

            remoteViews.setImageViewResource(R.id.enable, getImage(enable))
            remoteViews.setOnClickPendingIntent(R.id.enable, EnableWidget.buildButtonPendingIntent(context))

            EnableWidget.pushWidgetUpdate(context, remoteViews)
        }

        fun getImage(enabled: Boolean): Int {
            return if (enabled) {
                R.drawable.widget_enabled
            } else {
                R.drawable.widget_disabled
            }
        }
    }
}
